zerver/tests/test_auth_backends.py

```diff
@@ -1872,6 +1872,364 @@ def test_idp_initiated_signin_subdomain_implicit_invalid(self) -> None:
+
+        self.assertEqual(result.status_code, 302)
+        confirmation = Confirmation.objects.all().last()
+        confirmation_key = confirmation.confirmation_key
+        self.assertIn('do_confirm/' + confirmation_key, result.url)
+        do_confirm_url = result.url
+        result = self.client_get(do_confirm_url, name='')
+        self.assert_in_response('action="/accounts/register/"', result)
+        confirmation_data = {"from_confirmation": "1",
+                             "key": confirmation_key}
+        result = self.client_post('/accounts/register/', confirmation_data)
+        self.assert_in_response("We just need you to do one last thing", result)
+
+        # Verify that the user is asked for name but not password
+        self.assert_not_in_success_response(['id_password'], result)
+        self.assert_in_success_response(['id_full_name'], result)
+
+        # Click confirm registration button.
+        result = self.client_post(
+            '/accounts/register/',
+            {'full_name': "Full Name",
+                'key': confirmation_key,
+                'terms': True})
+
+        self.assertEqual(result.status_code, 302)
+
+        user_profile = get_user_by_delivery_email(email, realm)
+        self.assert_logged_in_user_id(user_profile.id)
+        self.assertEqual(user_profile.full_name, "Full Name")
+
+        self.assertFalse(user_profile.has_usable_password())
+
+    def test_validate_state(self) -> None:
+        with mock.patch('zproject.backends.logging.info') as m:
+
+            # check if auth fails if no state value is sent.
+            result = self.client_post('/complete/apple/')
+            self.assertEqual(result.status_code, 302)
+            self.assertIn('login', result.url)
+            m.assert_called_with("Sign in with Apple failed: missing state parameter.")
+
+            # Check if auth fails when a state sent has no valid data stored in redis.
+            fake_state = "fa42e4ccdb630f0070c1daab70ad198d8786d4b639cd7a1b4db4d5a13c623060"
+            result = self.client_post('/complete/apple/', {'state': fake_state})
+            self.assertEqual(result.status_code, 302)
+            self.assertIn('login', result.url)
+            m.assert_called_with("Sign in with Apple failed: bad state token.")
+
+class AppleAuthBackendNativeFlowTest(AppleAuthMixin, SocialAuthBase):
+    '''
+    This class is to test the native flow we continue with the
+    data sent as GET request from mobile app.
+    '''
+
+    __unittest_skip__ = False
+
+    # This is to restrict `prepare_login_url_and_headers` from
+    # taking the flow to signup page when `is_signup=True`
+    SIGNUP_URL = '/complete/apple/'
+
+    # This is just to inhibit `super().prepare_login_url_and_headers`
+    # failing because of assigning `url=self.LOGIN_URL` which gets
+    # changed because of `alternative_start_url` we send as a param
+    # and has no role in any test.
+    LOGIN_URL = ''
```

Comment:
I don't understand this comment at all.

Reply:
seeing this code makes it clear  I think.
```
def prepare_login_url_and_headers(
        self,
        subdomain: Optional[str]=None,
        mobile_flow_otp: Optional[str]=None,
        desktop_flow_otp: Optional[str]=None,
        is_signup: bool=False,
        next: str='',
        multiuse_object_key: str='',
        alternative_start_url: Optional[str]=None,
        *,
        user_agent: Optional[str]=None,
    ) -> Tuple[str, Dict[str, Any]]:
        url = self.LOGIN_URL
        if alternative_start_url is not None:
            url = alternative_start_url

        params = {}
        headers = {}
        if subdomain is not None:
....
...
`
there's `url = self.LOGIN_URL` before url gets assigned to `alternative_start_url` in the `if` block. If `LOGIN_URL` is missed, that functions fails since it cannot find `LOGIN_URL`. 
As url is getting changed every time to `alternative_url` replacing `LOGIN_URL`, we say `LOGIN_URL` has no role in tests.
I'll try to make it a bit clear.

Reply:
Modified the comment. Hope it's clear now :slightly_smiling_face: